### DISCUSSION TITLE
Client-side routing accessibility

### DISCUSSION THREAD

**Discussion by rjackson** [vercel/next.js#28969] [Help]
Hi folks,

I have an accessibility problem with client-side routing that I've no idea how to deal with, with regards to screen readers' interpretation of page transitions.

To help me explain the problem, I have set up a demo project:

- Repository: <https://github.com/rjackson/demo-client-side-routing-a11y>
- Hosted: <https://demo-client-side-routing-a11y.netlify.app/>

The homepage of this demo project lists 4 groups of pages:

* Client-side routing, without live regions (on the page)
* Client-side routing, with live regions
* Full reload transitions, without live regions
* Full reload transitions, with live regions

Each group contains two pages, and each page mocks a departures board for Manchester Piccadilly and Manchester Victoria train stations.

I am testing the transitions between pages in each group by using VoiceOver on a Mac, and I am ignoring the initial transition to each group from the index page (that's a client-side transition).

**Expectations**

The way the "Full reload transitions" work (with and without live regions) are how I expect navigation to work across Next.js as a whole:

1. Starting on the _Manchester _Piccadilly_ page, use VoiceOver to navigate to the "Manchester Victoria" Link
2. Hit enter
3. There is a full page transition to the Manchester Victoria page
3. **VoiceOver then announces the page top to bottom:

```
Navigation.
Heading level 2, Full page transitions, ...
List 3 items
...
```

The fourth step of what VoiceOver announces is what I am paying attention to, and is what differs with client-side routing.

**Behaviour of client-side routing, without live regions**

Following the same steps as above, when I click on "Manchester Victoria" VoiceOver announces:

```
Manchester Victoria.
```

And when navigating back to "Manchester Piccadilly", it just reads "Manchester Piccadilly". VoiceOver does not reset and start reading from the top of the page.

**Behaviour of client-side routing, with live regions**

When there are live regions on a page, VoiceOver detects the changes to those live regions and announces the changes within them, instead of announcing the entire page from top to bottom.

This can be very confusing, and is how I realised this issue with client-side routing: I was wondering why I was being told new departure information instead of going to a new page.

So in these scenarios, when clicking "Manchester Victoria" VoiceOver announces: 

```
Manchester Victoria.
Departures.
Bullet. Bullet. Train to Bristol in 9 minutes.
Bullet. Bullet. Train to Liverpool in 14 minutes.
```

**Possible solution**

The following potential fix does appear to work, **but I must caveat it with the fact that I am not an accessibility expert**. Whilst it appears to resolve the issues I'm seeing with the exact scenarios I'm testing, and the exact way I am testing it, I do not know if it might have any implications to other accessibility tools & software.

```diff
-import '../styles/globals.css'
+import { useRouter } from "next/dist/client/router";
+import "../styles/globals.css";
 
 function MyApp({ Component, pageProps }) {
-  return <Component {...pageProps} />
+  const router = useRouter();
+
+  return (
+    <div aria-live="assertive" aria-atomic="true" key={router.asPath}>
+      <Component {...pageProps} />
+    </div>
+  );
 }
 
-export default MyApp
+export default MyApp;
```

The intent behind this solution is:

* Have screen readers announce that the page has changed, due to `aria-live` and `aria-atomic` attributes wrapping the page
* Force React to re-render all of the components within the page, so if the page contained any  live sections of its own they would be removed and re-added to the DOM – and thus not announce any changes within themselves.

This potential fix can be viewed online at <https://potential-fix--demo-client-side-routing-a11y.netlify.app/>

----

Through this discussion I'm hoping we can ascertain whether or not there is a problem with Next.js' client-side routing with regards to accessibility, whether there are any known solutions, or if not whether there are any problems with my attempted solution above.

Thank y'all for reading :)

(accidentally submitted this discussion before I'd finished writing it up, apologies!)

**Comment by shuding** [comment]
Thanks for sharing! If I understand correctly, this is your main concern:

> VoiceOver does not reset and start reading from the top of the page.

Yeah I believe that’s something that needs some more further discussions. I don’t think reading/focusing again from the top of the page is always the best behavior, sometimes we want to retain focus on the current element. For example [this article](https://www.smashingmagazine.com/2015/05/client-rendered-accessibility) mentions:

> If you’re focused on an item and it gets rerendered, do you bang your keyboard against the desk and start over at the top of the page or gracefully continue on your way?

In SPAs a route change doesn’t always mean a full page transition (for example, it can be a popup like the Instagram web app). Changing focus point aggressively _might_ result in bad experience, and it breaks behaviors like `autofocus` for non screen reader users too. Some prior discussions in Next.js can be found in #7681 and related PRs. Overall I think this is more like a UX decision that Next.js cannot make for you.

And I have a couple of questions regarding your attempted solution:
- Is it necessary to mark the entire page as an assertive live region? According to MDN “(assertive) should only be used for time-sensitive/critical notifications that absolutely require the user's immediate attention.”
- The key prop will cause the entire page being re-mounted, so all internal component states will be lost and that might be a problem.
- Will any change inside that page be announced due to `assertive`, even if it’s not a navigation?

**Reply by rjackson** [reply]
Thanks for your thoughts & questions @shuding! I've also been chatting about this with my colleagues and we've had some further thoughts on it as well.

To distill my thoughts, I have two main concerns:

1. When changing page, VoiceOver does not reset and start reading from the top of the page.
2. When `aria-live` attributes are within the page, a page change is triggering any `aria-live` elements ("Live regions") to announce their content changes

I believe we should dismiss that first concern, because its a speculative concern from myself – with a novice understanding of accessibility, simply trying to emulate what happens when there is a full page transition. I have no insight into what would be desirable to regular users of assistive technology, and Next.js' currently behaviour is as per Marcy Sutton's research with regular users of assistive tech ([What we learned from user testing of accessible client-side routing techniques with Fable Tech Labs
](https://www.gatsbyjs.com/blog/2019-07-11-user-testing-accessible-client-routing/#common-accessibility-barriers-in-client-rendered-web-apps)).

In dismissing that first concern, I'll also retract the "possible solution" proposed above as that was focused on the first concern. As you highlight, there are valid questions about that as a potential solution that would need exploring if we thought that first concern was valid.

The second concern is outstanding. Because React is computing the diff between the actual DOM and what it think the DOM should be, page transitions will change the contents of live regions and thus cause those changes to be announced.

A revised solution that my colleagues and I have come to is that instead of remounting the entire page (which I originally thought of, because I've needed to do this to solve #9992 in a different project), **we only need to remount the live region**.

I believe we could achieve by `key`-ing the live region with the current page route, although I've yet to test that.

If we come to an agreement that this second concern is valid, I wonder if there is value in Next.js publishing some guidance in the relevant documentation (page routing?) about this problem, and potential solutions – either the above, if we find it works, or other ideas we might find through further discussion.
